### PR TITLE
Revision blacklisting

### PR DESCRIPTION
**Pull request by boehmseb**
This PR introduces the possibility to mark single revisions or revision ranges of case studies as _blocked_.

Blocked revisions cannot be used sensibly in experiments for some reason (e.g. the revision contains a bug that prevents it from compiling).

As an example to demonstrate how blocked revisions could be handled by other tools of the TS, I added support for blocked revisions to `vara-cs status`.

Blocked revisions are currently applied to the following case studies (see se-passau/vara-paper-configs#1):
 
- [ ] doxygen
- [ ]  git
- [x]  gravity (partially)
- [ ]  gzip
- [ ]  libvpx
- [ ]  lrzip
- [ ]  tmux
- [ ]  vim
- [ ]  x264
- [ ]  xz

For a clean merge, I suggest to wait until #149 is merged as this PR depends on some of its functionality.

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Clones removed
==============
+ varats/projects/c_projects/gravity.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4463811&bid=15077903)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Clones removed
==============
+ varats/projects/c_projects/gravity.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4463811&bid=15077903)

File: varats/data/report.py [hunk 1]
```diff
@@ -151,6 +152,12 @@ def result_file_has_status_missing(file_name: str) -> bool:
         return MetaReport.result_file_has_status(file_name,
                                                  FileStatusExtension.Missing)
 
+    @staticmethod
+    def result_file_has_status_blocked(file_name: str) -> bool:
+        """ Check if the passed file name is a (Missing) result file. """
+        return MetaReport.result_file_has_status(file_name,
+                                                FileStatusExtension.Blocked)
```
Comment (vulder) [inline varats/data/report.py]:
```suggestion
                                                 FileStatusExtension.Blocked)
```

File: varats/projects/c_projects/gravity.py [hunk 2]
```diff
@@ -30,14 +35,41 @@ class Gravity(Project):  # type: ignore
...
+                "46133fb47d6da1f0dec27ae23db1d633bc72e9e3", "master")
+
+        if version_id in cmake_revisions:
+            print("CMAKE")
+            self.__compile_cmake()
+        else:
+            print("MAKE")
+            self.__compile_make()
+
+    def __compile_cmake(self):
```
Comment (vulder) [inline varats/projects/c_projects/gravity.py]: did you include your other work from PR #149  here too?

File: varats/utils/project_util.py [hunk 3]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+    """
+    Returns a list of all revisions between two commits a and b (inclusive),
+    where a comes before b.
+    It is assumed that the current working directory is the git repository.
+    """
+    result = [a]
+    result.extend(
+        git("log", "--pretty=%H", "--ancestry-path",
+            "{}..{}".format(a, b)).strip().split())
+    return result
```
Comment (vulder) [inline varats/utils/project_util.py]: same #149

File: varats/projects/c_projects/gravity.py [hunk 4]
```diff
@@ -30,14 +35,41 @@ class Gravity(Project):  # type: ignore
...
         with local.cwd(self.SRC_FILE):
             with local.env(CC=str(clang)):
                 cmake("-G", "Unix Makefiles", ".")
             run(make["-j", int(CFG["jobs"])])
+
+    def __compile_make(self):
+        clang = cc(self)
+        with local.cwd(self.SRC_FILE):
+            with local.env(CC=str(clang)):
+                run(make["-j", int(CFG["jobs"])])
```
Comment (vulder) [inline varats/projects/c_projects/gravity.py]: same #149

File: varats/utils/project_util.py [hunk 5]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+
+
+class BlockedRevision():
+    """
+    A revision marked as blocked due to some `reason`.
+    """
+
+    def __init__(self, id: str, reason: tp.Optional[str] = None):
```
Comment (vulder) [inline varats/utils/project_util.py]: `id` is a python builtin and naming the var like this redefines this builtin

File: varats/utils/project_util.py [hunk 6]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+        self.__id = id
+        self.__reason = reason
+
+    @property
+    def reason(self):
```
Comment (vulder) [inline varats/utils/project_util.py]: should have a short comment to describe the property

File: varats/utils/project_util.py [hunk 7]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+                 id_end: str,
+                 reason: tp.Optional[str] = None):
+        self.__id_start = id_start
+        self.__id_end = id_end
+        self.__reason = reason
+        # cache for commit hashes
+        self.__revision_list: tp.Optional[tp.List[str]] = None
+
+    @property
+    def reason(self):
```
Comment (vulder) [inline varats/utils/project_util.py]: same

File: varats/utils/project_util.py [hunk 8]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+        if self.__revision_list is None:
+            self.__revision_list = get_all_revisions_between(
+                self.__id_start, self.__id_end)
+
+        return self.__revision_list.__iter__()
+
+
+class BlockedRevisionCheckerDelegate():
+    """
+    Delegate for the
```
Comment (vulder) [inline varats/utils/project_util.py]: for the ???

File: varats/utils/project_util.py [hunk 9]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+    """
+
+    def __init__(self, project_name: str) -> None:
+        self.__project_name = project_name
+        self.__project_path = get_local_project_git_path(project_name)
+
+        self.__blacklist_entries: tp.List[
+            tp.Union[BlockedRevision, BlockedRevisionRange]] = []
+
+    def is_blocked_revision(self, id: str) -> tp.Tuple[bool, tp.Optional[str]]:
```
Comment (vulder) [inline varats/utils/project_util.py]: missing comment

File: varats/utils/project_util.py [hunk 10]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+        # cd to repo because of potential git lookups
+        with local.cwd(self.__project_path):
+            for b_entry in self.__blacklist_entries:
+                for b_item in b_entry:
+                    if b_item.startswith(id):
+                        return True, b_entry.reason
+        return False, None
+
+    def block_revision(self, id: str, reason: tp.Optional[str] = None) -> None:
```
Comment (vulder) [inline varats/utils/project_util.py]: missing comment and `id`

File: varats/utils/project_util.py [hunk 11]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+        """
+        self.__blacklist_entries.append(
+            BlockedRevisionRange(id_start, id_end, reason))
+
+
+class BlockedRevisionChecker():
+    """
+    Interface for blacklisting/blocking revisions from a project.
+
+    Implementors must initialize the delegate with a call to
```
Comment (vulder) [inline varats/utils/project_util.py]:
```suggestion
    Implementors must initialize the delegate with a call to
```

File: varats/utils/project_util.py [hunk 12]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+    """
+
+    __blocked_revision_checker_delegate: tp.Optional[
+        BlockedRevisionCheckerDelegate] = None
+
+    @classmethod
+    def initialize_for_project(cls, project: str):
+        """
+        Initializes the revision checker delegate.
+
```
Comment (vulder) [inline varats/utils/project_util.py]:
```suggestion

```

File: varats/utils/project_util.py [hunk 13]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+        """
+        Initializes the revision checker delegate.
+        
+        This needs to be called before revisions can be blocked.
+        """
+        cls.__blocked_revision_checker_delegate = \
+            BlockedRevisionCheckerDelegate(project)
+
+    @classmethod
+    def block_revision(cls, id: str, reason: tp.Optional[str] = None) -> None:
```
Comment (vulder) [inline varats/utils/project_util.py]: `id`

File: varats/utils/project_util.py [hunk 14]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+        Blacklist all revisions between commit `id_start` (older)
+        and `id_end` (newer).
+        """
+        if cls.__blocked_revision_checker_delegate is None:
+            return
+        cls.__blocked_revision_checker_delegate.block_revisions(
+            id_start, id_end, reason)
+
+    @classmethod
+    def is_blocked_revision(cls, id: str) -> tp.Tuple[bool, tp.Optional[str]]:
```
Comment (vulder) [inline varats/utils/project_util.py]: `id`

File: varats/utils/project_util.py [hunk 15]
```diff
@@ -72,3 +72,153 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
...
+    Interface for blacklisting/blocking revisions from a project.
+
+    Implementors must initialize the delegate with a call to 
+    `initialize_for_project()` before revisions can be blacklisted.
+    This is necessary because we need to know which project we are dealing
+    with in order to compute revision ranges.
+    """
+
+    __blocked_revision_checker_delegate: tp.Optional[
+        BlockedRevisionCheckerDelegate] = None
```
Comment (vulder) [inline varats/utils/project_util.py]: Isn't this static method overwritten when another project initializes a BRChecker?

**Review by vulder — CHANGES_REQUESTED** [review]
This should be rebased after merging #149

File: varats/projects/c_projects/gravity.py (hunk 2 above)
Comment (boehmseb) [reply varats/projects/c_projects/gravity.py]: No, but this was in the same commit as the needed `get_all_revisions_between`.

File: varats/utils/project_util.py (hunk 15 above)
Comment (boehmseb) [reply varats/utils/project_util.py]: You are right. I thought that this call would act as if it was called via `super()`, but it does not.
Even worse: this cannot be done at all inside the (sub-) class definition, as the corresponding class object does not even exist at that point.
A decorator might help in this case. I will look into that later.

**Comment by boehmseb** [comment]
I rewrote the revision blocking code to be a decorator. This is even easier to use, needs less code and resolves the problem stated above.

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones removed
==============
+ varats/projects/c_projects/gravity.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4463811&bid=15077903)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones removed
==============
+ varats/projects/c_projects/gravity.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4463811&bid=15077903)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones removed
==============
+ varats/projects/c_projects/gravity.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4463811&bid=15077903)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ varats/projects/c_projects/gravity.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4463811&bid=15077903)

**Comment by boehmseb** [comment]
The current CI fail is caused by mypy and will be gone after rebasing onto PR #149.

**Review by vulder — APPROVED** [review]
lgtm

**Pull request merged**